### DISCUSSION TITLE
What type should ref be

### DISCUSSION THREAD

**Discussion by benesva4** [react-hook-form/react-hook-form#1950] [Q&A] [answered]
Hello everyone I need a little help please.
I have my own component, it's basically just a wrapper for the label and an input with some logic. I want to pass a register to the components, so I can use it on the nested input, but I am not sure what type should I give it. It even seems, like the type of the register depends on the values of the form, so maybe it could end up in some kind of generics hell.

Simplified version looks something like this: 
```
type Props = {
  register: typeOfRef; <------------------ THIS ONE
  label: string;
  name: string;
}

const MyInput = ({ label, register, name }: Props) => (
  <div>
    <label for={name}>{label}</label>
    <input id={name} name={name} ref={register()} />
  </div>
);

const { register } = useForm()
...
<MyInput name="firstName" label="First name" ref={register} />
<MyInput name="lastName" label="Last name" ref={register} />
``` 
Maybe there is different way to do this I don't know.

Thank you very much for any suggestion. :-)

P.S. For the creators - Thanks for this awesome library.

**Comment by keiya01** [comment, marked as answer]
I think you can specify `forwardRef` in React. Using `forwardRef`, you can pass `ref` props.

For example:

```tsx

const Input = React.forwardRef<HTMLInputElement, { name: string, ...etc }>((props, ref) => <input {...props} ref={ref}/>);

```

https://reactjs.org/docs/forwarding-refs.html

**Reply by benesva4** [reply]
Thank you very much. I thought it would be more on the React side...

**Reply by benesva4** [reply]
By the way, could this be the solution too?
https://react-hook-form.com/api#useFormContext

**Reply by keiya01** [reply]
Why could you not use `forwardRef`?
I think `useFormContext` should be used when you want to have state between many component. Or you should be used when you want to save some entered value by user at temporary.

**Reply by benesva4** [reply]
I asked more like out of curiosity, I used that ref solution.
